org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -573,12 +576,39 @@ protected Compiler newCompiler() {
 	CompilerOptions compilerOptions = new CompilerOptions(projectOptions);
 	compilerOptions.performMethodsFullRecovery = true;
 	compilerOptions.performStatementsRecovery = true;
-	Compiler newCompiler = new Compiler(
-		this.nameEnvironment,
-		DefaultErrorHandlingPolicies.proceedWithAllProblems(),
-		compilerOptions,
-		this,
-		ProblemFactory.getProblemFactory(Locale.getDefault()));
+	Compiler newCompiler = null;
+	String compilerClassName = System.getProperty(AbstractImageBuilder.class.getSimpleName() + ".compiler"); //$NON-NLS-1$
+	if (compilerClassName != null) {
+		try {
+			Class<? extends Compiler> compilerClass = (Class<? extends Compiler>) Class.forName(compilerClassName);
+			Constructor<? extends Compiler> constructor = compilerClass.getDeclaredConstructor(
+				INameEnvironment.class,
+				IErrorHandlingPolicy.class,
+				CompilerConfiguration.class,
+				ICompilerRequestor.class,
+				IProblemFactory.class);
+			newCompiler = constructor.newInstance(this.nameEnvironment,
+				DefaultErrorHandlingPolicies.proceedWithAllProblems(),
+				prepareCompilerConfiguration(compilerOptions),
+				this,
+				ProblemFactory.getProblemFactory(Locale.getDefault()));
+			this.useDefaultCompiler = false;
```

Comment:
It's not good when the method newCompiler() changes the state of the Class, as successive calls could alter the behaviour. It would be better a flag Compiler.isDefaultCompiler. But ideally there would not be such flag at all but participant.isAnnotationProcessor() or participant.processAnnotations() would just behave different.

Reply:
useDefaultCompiler is removed.